File: examples/haskell/HelloWorld.hs
module Main where

import Azul
import qualified Azul.App as App
import qualified Azul.AppConfig as AppConfig
import qualified Azul.Button as Button
import qualified Azul.Dom as Dom
import qualified Azul.WindowCreateOptions as WindowCreateOptions

newtype DataModel = DataModel { counter :: Int }

onClick :: DataModel -> CallbackInfo -> (DataModel, Update)
onClick model _ = (model { counter = counter model + 1 }, Update_RefreshDom)

layout :: DataModel -> LayoutCallbackInfo -> IO Dom
layout model _ = do
  label <- Dom.createPWithText (show (counter model)) >>= Dom.withCss "font-size: 32px; margin: 0;"
  button <- Button.create "Increase counter"
    >>= Button.withButtonType ButtonType_Primary
    >>= Button.onClick onClick
    >>= Button.dom
  Dom.createBody >>= Dom.withChild label >>= Dom.withChild button

main :: IO ()
main = do
  window <- WindowCreateOptions.create layout
  AppConfig.create >>= App.create (DataModel 5) >>= App.run window

File: examples/csharp/hello-world.cs
using System;
using Azul;

namespace HelloWorld
{
    public sealed class MyDataModel
    {
        public uint Counter;
        public MyDataModel(uint counter) { Counter = counter; }
    }

    public static class Program
    {
        private static readonly MyDataModel _model = new MyDataModel(5);

        private static Update OnClick(MyDataModel m, IntPtr info)
        {
            m.Counter += 1;
            return Update.RefreshDom;
        }

        private static Dom Layout(MyDataModel m, IntPtr info)
        {
            var label = Dom.CreatePWithText(m.Counter.ToString())
                .WithCss("font-size: 32px; margin: 0;");
            var buttonDom = Button.Create("Increase counter")
                .WithButtonType(ButtonType.Primary)
                .OnClick(m, OnClick)
                .Dom();
            return Dom.CreateBody()
                .WithChild(label)
                .WithChild(buttonDom);
        }

        public static int Main(string[] args)
        {
            using var app = App.Create(_model, AppConfig.Create());
            app.Run(WindowCreateOptions.Create<MyDataModel>(Layout));
            return 0;
        }
    }
}
